File: 1524.number-of-sub-arrays-with-odd-sum.py
#
# @lc app=leetcode id=1524 lang=python3
#
# [1524] Number of Sub-arrays With Odd Sum
#

# @lc code=start
class Solution:
    def numOfSubarrays(self, arr: list[int]) -> int:
        ans = 0
        even = 1 #(sum zero before array start)
        odd = 0
        rsum = 0
        for i in range(len(arr)):
            rsum += arr[i]
            if rsum % 2 == 1:
                ans += even
                odd += 1
            else:
                ans += odd
                even += 1
        return ans % (10**9 + 7)
        
# @lc code=end

File: 1524.number-of-sub-arrays-with-odd-sum.java
/*
 * @lc app=leetcode id=1524 lang=java
 *
 * [1524] Number of Sub-arrays With Odd Sum
 */

// @lc code=start
class Solution {
    public int numOfSubarrays(int[] arr) {
        //Rule: odd sum - odd sum = even sum
        //Rule: even sum - even sum = even sum
        //Rule: odd sum - even sum = odd sum
        //Rule: even sum - odd sum = odd sum
        int odd = 0, even = 1, sum = 0;
        long ans = 0;
        for(int num : arr ){
            sum += num;
            if(sum % 2 == 0){
                //prevent overflow problem
                ans = (ans + odd) % 1000000007;
                even++;
            }
            else{
                ans = (ans + even) % 1000000007;
                odd++;
            }
        }
        return (int) ans;
    }
}
// @lc code=end

